File: 2021/02/main.jl
function parsecmd(line)
    words = split(line)
    return (words[1], parse(Int, words[2]))
end

function move(pos, direction, steps)
    direction == "up" && return (pos[1], pos[2] - steps)
    direction == "down" && return (pos[1], pos[2] + steps)
    direction == "forward" && return (pos[1] + steps, pos[2])
end

function move_with_aim(pos, direction, steps)
    direction == "up" && return (pos[1], pos[2], pos[3] - steps)
    direction == "down" && return (pos[1], pos[2], pos[3] + steps)
    direction == "forward" && return (pos[1] + steps, pos[2] + steps * pos[3], pos[3])
end

function main(input_path)
    commands = parsecmd.(eachline(input_path))
    pos = (0, 0)
    for cmd in commands
        pos = move(pos, cmd...)
    end
    println(pos[1] * pos[2])
    pos = (0, 0, 0)
    for cmd in commands
        pos = move_with_aim(pos, cmd...)
    end
    println(pos[1] * pos[2])
end

if abspath(PROGRAM_FILE) == @__FILE__
    main(joinpath(dirname(@__FILE__), "input"))
end

File: 2021/02/main.py
import pathlib

def parse(line):
    direction, steps = line.split(" ")
    return direction, int(steps)

def move(pos, direction, steps):
    if direction == "up":
        return (pos[0], pos[1] - steps)
    if direction == "down":
        return (pos[0], pos[1] + steps)
    if direction == "forward":
        return (pos[0] + steps, pos[1])
    raise ValueError

def move_with_aim(pos, direction, steps):
    if direction == "up":
        return (pos[0], pos[1], pos[2] - steps)
    if direction == "down":
        return (pos[0], pos[1], pos[2] + steps)
    if direction == "forward":
        return (pos[0] + steps, pos[1] + steps * pos[2], pos[2])
    raise ValueError

def main(input_path):
    with open(input_path) as fp:
        commands = list(map(parse, fp))
    pos = (0, 0)
    for cmd in commands:
        pos = move(pos, *cmd)
    print(pos[0] * pos[1])
    pos = (0, 0, 0)
    for cmd in commands:
        pos = move_with_aim(pos, *cmd)
    print(pos[0] * pos[1])

if __name__ == "__main__":
    main(pathlib.Path(__file__).parent.resolve() / "input")
